### PR TITLE
Update .NET SDK to Preview 2

### PR DESCRIPTION
**Pull request by mmitche**
### To double check:

* [ ] The right tests are in and and the right validation has happened.  Guidance:  https://github.com/dotnet/arcade/tree/main/Documentation/Validation

### CODE REVIEW

**Comment by mmitche** [comment]
@carlossanlop @hoyosjs Does that core dump when running p2 on ubuntu-latest look familiar at all?

```
Welcome to .NET 7.0!
---------------------
SDK Version: 7.0.100-preview.2.22153.17

----------------
Installed an ASP.NET Core HTTPS development certificate.
To trust the certificate run 'dotnet dev-certs https --trust' (Windows and macOS only).
Learn about HTTPS: https://aka.ms/dotnet-https
----------------
Write your first app: https://aka.ms/dotnet-hello-world
Find out what's new: https://aka.ms/dotnet-whats-new
Explore documentation: https://aka.ms/dotnet-docs
Report issues and find source on GitHub: https://github.com/dotnet/core
Use 'dotnet --help' to see available commands or visit: https://aka.ms/dotnet-cli
--------------------------------------------------------------------------------------
Cannot get required symbol SSL_set_alpn_protos from libssl
/__w/1/s/eng/common/tools.sh: line 474:   488 Aborted                 (core dumped) "$_InitializeBuildTool" "$@"
Build failed with exit code 134. Check errors above.
Killing running build processes...
```

**Comment by hoyosjs** [comment]
Not really. Also @bartonjs, given the missing libssl symbol... It's 20.04 with OpenSSL 1.1.1f, which I expect to have said symbols...

**Comment by hoyosjs** [comment]
I also can't repro with openssl 1.1.1f-1ubuntu2.12 amd64 on 20.04:

```
root@5c2c8e61a500:/# readelf --dyn-syms /lib/x86_64-linux-gnu/libssl.so.1.1 | grep SSL_set_alpn_protos
   907: 000000000003a100   147 FUNC    GLOBAL DEFAULT   15 SSL_set_alpn_protos@@OPENSSL_1_1_0
```

That missing method would definitely cause that... but seems like a custom build of libssl in the agent? Is there a way to get a hold of the image?

**Comment by mmitche** [comment]
> I also can't repro with openssl 1.1.1f-1ubuntu2.12 amd64 on 20.04:
> 
> ```
> root@5c2c8e61a500:/# readelf --dyn-syms /lib/x86_64-linux-gnu/libssl.so.1.1 | grep SSL_set_alpn_protos
>    907: 000000000003a100   147 FUNC    GLOBAL DEFAULT   15 SSL_set_alpn_protos@@OPENSSL_1_1_0
> ```
> 
> That missing method would definitely cause that... but seems like a custom build of libssl in the agent? Is there a way to get a hold of the image?

Interesting. @MattGal @dotnet/dnceng, I know we had a set of images created that mirror the hosted pool images. Is there one for ubuntu-latest or whatever that maps too?

**Comment by MattGal** [comment]
> > I also can't repro with openssl 1.1.1f-1ubuntu2.12 amd64 on 20.04:
> > ```
> > root@5c2c8e61a500:/# readelf --dyn-syms /lib/x86_64-linux-gnu/libssl.so.1.1 | grep SSL_set_alpn_protos
> >    907: 000000000003a100   147 FUNC    GLOBAL DEFAULT   15 SSL_set_alpn_protos@@OPENSSL_1_1_0
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > That missing method would definitely cause that... but seems like a custom build of libssl in the agent? Is there a way to get a hold of the image?
> 
> Interesting. @MattGal @dotnet/dnceng, I know we had a set of images created that mirror the hosted pool images. Is there one for ubuntu-latest or whatever that maps too?

I didn't make a "latest" tag though that's possible if we think it's really valuable.  You can see them at the bottom of https://helix.dot.net/.  20.04 is currently the same as latest per https://github.com/actions/virtual-environments. 

You want `1es-ubuntu-2004-open` for the public project, and `1es-ubuntu-2004` for internal.  Ping me if you hit any trouble; we've already got a fair bit of builds moved over here.

**Comment by am11** [comment]
We are using Ubuntu 14:

image

which is out of support in .NET 6: https://github.com/dotnet/core/blob/main/release-notes/6.0/supported-os.md#linux

I think we should be using 18.04 or 20.04, to match runtime pipelines?

**Comment by am11** [comment]
If the build is failing inside Ubuntu 14, that will explain the reason of missing symbol. See this workaround for Ubuntu 14: https://gist.github.com/am11/11b8f526747c5ad12c9a74e7765e14a4#file-dockerfile-L12. We already deleted trusty setup in https://github.com/dotnet/arcade/pull/7694, just need to update container references in pipelines to use something more recent (i.e. one of the [supported version of Ubuntu](https://github.com/dotnet/core/blob/main/release-notes/6.0/supported-os.md#linux)).

**Comment by mmitche** [comment]
The SDK itself no longer supports 14.04, so running inside the 14.04 container wouldn't necessarily be expected to work. I've updated to the cotnainer the runtime uses for linux x64

**Comment by hoyosjs** [comment]
Actually - we should just update. Arcade has no native deps and it's using `ubuntu-14.04-cross-0cd4667-20170319080304` (I somehow managed to overlook it was containerized). That's a cross-build image and it's not needed here.

**Comment by mmitche** [comment]
@MattGal This looks a bit puzzling. It's looking for aspnetcore but using the version of the core runtime. I think this is because this line https://github.com/dotnet/arcade/pull/8624/files#diff-8ab8c282f3d9822870eee0d25497d5ed8a335820fe03b6194b2dce937e1ca321L9 has an empty `NetCoreSdkVersion` property.

Does this look familiar or should I dig further?

**Comment by MattGal** [comment]
> @MattGal This looks a bit puzzling. It's looking for aspnetcore but using the version of the core runtime. I think this is because this line https://github.com/dotnet/arcade/pull/8624/files#diff-8ab8c282f3d9822870eee0d25497d5ed8a335820fe03b6194b2dce937e1ca321L9 has an empty `NetCoreSdkVersion` property.
> 
> Does this look familiar or should I dig further?

Not instantly familiar but I'm taking a peek

File: src/Microsoft.DotNet.Helix/Sdk/tools/dotnet-cli/DotNetCli.props [hunk 1]
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup>
     <IncludeDotNetCli Condition=" '$(IncludeDotNetCli)' != 'true' ">false</IncludeDotNetCli>
     <!-- TODO (https://github.com/dotnet/arcade/issues/7787): When bumping this, we should do #7787 -->
-    <AspNetCoreRuntimeVersion>6.0.0</AspNetCoreRuntimeVersion>
+    <AspNetCoreRuntimeVersion>7.0.0-preview.2.22153.2</AspNetCoreRuntimeVersion>
```
Comment (MattGal) [inline src/Microsoft.DotNet.Helix/Sdk/tools/dotnet-cli/DotNetCli.props:5]: This isn't meant to be a preview / unstable version, rather it's what Xharness expects.  I don't think you should rev the version here until 7.0.0 is GA

File: src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets [hunk 2]
```diff
@@ -4,7 +4,7 @@
     <IncludeDotNetCli>true</IncludeDotNetCli>
     <IncludeXHarnessCli>true</IncludeXHarnessCli>
     <XHarnessTargetFramework Condition=" '$(XHarnessTargetFramework)' == '' ">net6.0</XHarnessTargetFramework>
-    <DotNetCliVersion Condition=" '$(DotNetCliVersion)' == '' ">6.0.100</DotNetCliVersion>
+    <DotNetCliVersion Condition=" '$(DotNetCliVersion)' == '' ">7.0.100-preview.2.22153.17</DotNetCliVersion>
```
Comment (MattGal) [inline src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets:7]: Same comment here... this is specific to Xharness, it doesn't need to be on the preview bleeding edge.

**Comment by MattGal** [comment]
@mmitche I think you should just revert the lines from the Xharness Helix SDK, and that what you're going for (assuming that build # is correct) will be brought in when https://github.com/dotnet/arcade/issues/8336 is done.

**Comment by mmitche** [comment]
@MattGal No dice on reverting the xharness SDK (either both `DotNetCli.props` and `XHarnessRunner.targets` or only `XHarnessRunner.targets`.

It appears that the SDK knows the runtime tied to the SDK in the global.json, and that's ending up as DotNetCliVersion.

**Comment by MattGal** [comment]
> @MattGal No dice on reverting the xharness SDK (either both `DotNetCli.props` and `XHarnessRunner.targets` or only `XHarnessRunner.targets`.
> 
> It appears that the SDK knows the runtime tied to the SDK in the global.json, and that's ending up as DotNetCliVersion.

Rats, I thought we honored the version in the .targets.  It might just be that you need the other work coming in from @chcosta , but I defer to @premun for the XHarness Xpertise

**Comment by premun** [comment]
XHarness flows are in no way special really. They just utilize the `IncludeDotNetCli` property like any client can do. The only way it is special is probably that it wants `aspnetcore-runtime` and no other flows really do that so it is catching an edge case where the versions of the runtime do not really match with what's coming from `global.json`.

There should also be some unit test somewhere testing that the `aspnetcore-runtime` is fine (@jonfortescue wrote it I believe). You can also check how we were doing these bumps before:
https://github.com/dotnet/arcade/commit/9fc83fbf76339c9dec4f5a8bf4718df2d42992a3

We can also use the full SDK with XHarness and things will work as well (seems to be hinted in #7787). We can try taking this:
https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets#L2-L10
and reducing it to this:
```xml
  <PropertyGroup Condition=" '$(IncludeXHarnessCli)' == 'true' ">
    <IncludeDotNetCli>true</IncludeDotNetCli>
    <XHarnessTargetFramework Condition=" '$(XHarnessTargetFramework)' == '' ">net6.0</XHarnessTargetFramework>
    <DotNetCliVersion Condition=" '$(DotNetCliVersion)' == '' ">6.0.100</DotNetCliVersion>
    <DotNetCliPackageType Condition=" '$(DotNetCliPackageType)' != 'sdk' ">sdk</DotNetCliPackageType>
  </PropertyGroup>
```

We can then bump `DotNetCliVersion` once XHarness consumes this and is built with .NET 7 preview 1. However, if we could get this version from `global.json`, we can get rid of this need for manual bumps. I know it's available in some MSBuild property somewhere.

File: src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets [hunk 3]
```diff
@@ -1,12 +1,9 @@
 <Project>
   <PropertyGroup Condition=" '$(IncludeXHarnessCli)' == 'true' ">
-    <!-- We then invoke the CLI DLL directly using .NET runtime without the need for .NET SDK -->
     <IncludeDotNetCli>true</IncludeDotNetCli>
-    <IncludeXHarnessCli>true</IncludeXHarnessCli>
     <XHarnessTargetFramework Condition=" '$(XHarnessTargetFramework)' == '' ">net6.0</XHarnessTargetFramework>
     <DotNetCliVersion Condition=" '$(DotNetCliVersion)' == '' ">6.0.100</DotNetCliVersion>
-    <DotNetCliPackageType Condition=" '$(DotNetCliPackageType)' != 'sdk' ">aspnetcore-runtime</DotNetCliPackageType>
-    <DotNetCliVersion Condition=" '$(DotNetCliVersion)' == '$(NETCoreSdkVersion)' AND '$(DotNetCliPackageType)' == 'aspnetcore-runtime' ">$(AspNetCoreRuntimeVersion)</DotNetCliVersion>
+    <DotNetCliPackageType Condition=" '$(DotNetCliPackageType)' != 'sdk' ">sdk</DotNetCliPackageType>
```
Comment (premun) [inline src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets:6]: Looking at it again, this is equivalent:
```suggestion
    <DotNetCliPackageType>sdk</DotNetCliPackageType>
```

**Comment by mmitche** [comment]
@premun No dice :/

**Comment by mmitche** [comment]
> > Unable to find dotnet cli sdk version **7.0.0-preview.2.22152.2** from any of the specified feeds.
> 
> Where does that number come from?

That's the dotnet runtime that is within the SDK specified in global.json

**Comment by premun** [comment]
image

Is this it?

I think this boils down again to our `FindDotNetVersion` target (or how it's called) that is getting the info from some json somewhere and it is why we needed to type down this ASP.NET Core runtime version in the first place. I remember a discussion here: https://github.com/dotnet/arcade/issues/7022

**Comment by mmitche** [comment]
@premun `DotNetCliVersion` is set here. I think maybe the right thing to do is to set it explicitly with 6.0.100, regardless of whethe it's already set.

**Comment by premun** [comment]
> @premun `DotNetCliVersion` is set here. I think maybe the right thing to do is to set it explicitly with 6.0.100, regardless of whethe it's already set.

@mmitche I am afraid your last solution won't work once XHarness consumes newest Arcade, it will be built with .NET 7 and start requiring .NET 7 runtime/sdk to be run

**Comment by mmitche** [comment]
> > That's the dotnet runtime that is within the SDK specified in global.json
> 
> Why does it want the runtime version when we specify SDK?

In DotNetCli.props:

```
<DotNetCliPackageType Condition=" '$(DotNetCliPackageType)' == '' ">runtime</DotNetCliPackageType>
    <DotNetCliVersion Condition=" '$(DotNetCliVersion)' == '' AND '$(DotNetCliPackageType)' == 'runtime' ">$(BundledNETCoreAppPackageVersion)</DotNetCliVersion>
    <!-- TODO (https://github.com/dotnet/arcade/issues/7022): We are hardcoding this version to use the one tied to the SDK version from global.json -->
    <DotNetCliVersion Condition=" '$(DotNetCliVersion)' == '' AND '$(DotNetCliPackageType)' == 'aspnetcore-runtime' ">$(AspNetCoreRuntimeVersion)</DotNetCliVersion>
    <DotNetCliVersion Condition=" '$(DotNetCliVersion)' == '' AND '$(DotNetCliPackageType)' == 'sdk' ">$(NETCoreSdkVersion)</DotNetCliVersion>
    <DotNetCliChannel Condition=" '$(DotNetCliChannel)' == '' ">Current</DotNetCliChannel>
```

**Comment by mmitche** [comment]
> > @premun `DotNetCliVersion` is set here. I think maybe the right thing to do is to set it explicitly with 6.0.100, regardless of whethe it's already set.
> 
> @mmitche I am afraid your last solution won't work once XHarness consumes newest Arcade, it will be built with .NET 7 and start requiring .NET 7 runtime/sdk to be run

When is this slated to happen? Maybe we just wait till that flows into here and then we update?

**Comment by premun** [comment]
> > @mmitche I am afraid your last solution won't work once XHarness consumes newest Arcade, it will be built with .NET 7 and start requiring .NET 7 runtime/sdk to be run
> 
> When is this slated to happen? Maybe we just wait till that flows into here and then we update?

We can totally do it in sync, this is not a big deal at all. I am more worried about the non-XHarness usages of Helix SDK which will probably still suffer from this behaviour when trying to install the SDK because something is probably broken there?

Feel free to merge this, I can do the XHarness bump tomorrow and bump it here as well.

**Review by premun — APPROVED** [review]

**Comment by riarenas** [comment]
@mmitche fyi that this is stuck in arcade-validation: https://github.com/dotnet/arcade-validation/pull/3022

I'm going to try and .net 7 all the things there real quick and if it's not straightforward I'll have FR take a look

**Comment by riarenas** [comment]
It's unblocked now
